slider/beatmap.py

```diff
@@ -909,6 +909,43 @@ def _get_as_bool(groups, section, field, default=no_default):
         )
 
 
+def _moving_average_by_time(data, delta, num):
+    """Performs a moving average over a time window of an array where the
+    first column is the time and the second the value.
+    Returns an array with the first column containing times and the second
+    containing the average at that time.
+    """
+    current_sum = 0
+    front = 0
+    back = 0
+    out = np.empty((num, 2))
+    for i, t in enumerate(np.linspace(0, data[-1][0], num)):
```

Comment:
One piece of advice is that looping of numpy arrays is usually indicative of numpy misuse. Numpy gets its speed by pushing loops in C and fortran which are much much faster than Python for iteration, using a loop over an array defeats this.

That said, this algorithm is quite tricky to implement more efficiently with numpy functions. I think the way to do this is with a somewhat obscure feature of ufuncs called `reduceat` which allows you to reduce over a sequence of slices. In this case, we can use this to sum up all of the windows around our sample points and then divide out the size of the window.

The full implementation for this looks like:
```python
def _moving_average_by_time(data, delta, num):
    """Take the moving average of some values and sample it at regular
    frequencies.

    Parameters
    ----------
    data : np.ndarray
        The array to reduce. This array should have two columns: the time
        in seconds and the value to take the average of.
    delta : int
        The length of the leading and trailing window.
    num : int
        The number of samples to take.

    Returns
    -------
    averages : np.ndarray
        A two column array where the first column is the time and the second
        column is the reduction at the time. The length of axis 0 is ``num``.
    """
    times = data[:, 0]

    # take an even sample from 0 to the end time
    out_times = np.linspace(0, times[-1], num)

    # compute the start and stop indices for each sampled window
    window_start_ixs = np.searchsorted(times, out_times - delta)
    window_stop_ixs = np.searchsorted(times, out_times + delta)

    # a 2d array of shape ``(num, 2)`` where each row holds the start and stop
    # index for the window
    window_ixs = np.stack([window_start_ixs, window_stop_ixs], axis=1)

    # append a nan to the end of the values so that we can do many slices all
    # the way to the end in reduceat
    values = np.append(data[:, 1], [np.nan])

    # sum the values in the ranges ``[window_start_ixs, window_stop_ixs)``
    window_sums = np.add.reduceat(values, window_ixs.ravel())[::2]
    window_sizes = np.diff(window_ixs, axis=1).ravel()
    # convert window_sizes of 0 to 1 (inplace) to prevent division by zero
    np.clip(window_sizes, 1, None, out=window_sizes)

    out_values = window_sums / window_sizes

    return np.stack([out_times, out_values], axis=1)
```

This runs considerably faster for even TV size songs which helps a lot if you are processing thousands of beatmaps or just want to improve latency when working with a single map.